File: sdks/php/TigerExClient.php
<?php
/**
 * TigerEx PHP SDK
 * Production-grade PHP SDK for TigerEx exchange API
 */

namespace TigerEx\SDK;

class TigerExClient
{
    private $apiKey;
    private $apiSecret;
    private $baseUrl;
    private $timeout;
    private $testnet;

    public function __construct(
        string $apiKey = '',
        string $apiSecret = '',
        bool $testnet = false,
        int $timeout = 30000
    ) {
        $this->apiKey = $apiKey;
        $this->apiSecret = $apiSecret;
        $this->testnet = $testnet;
        $this->baseUrl = $testnet 
            ? 'https://api-test.tigerex.com' 
            : 'https://api.tigerex.com';
        $this->timeout = $timeout;
    }

    // =========================================================================
    // PRIVATE METHODS
    // =========================================================================

    private function sign(string $message): string
    {
        return hash_hmac('sha256', $message, $this->apiSecret);
    }

    private function buildQueryString(array $params): string
    {
        ksort($params);
        $query = http_build_query($params, '', '&');
        return $query;
    }

    private function request(
        string $method,
        string $endpoint,
        array $params = [],
        bool $signed = false
    ): array {
        $timestamp = round(microtime(true) * 1000);
        $queryString = $this->buildQueryString($params);

        $url = $this->baseUrl . $endpoint;
        if ($method === 'GET' && !empty($queryString)) {
            $url .= '?' . $queryString;
        }

        $headers = [
            'Content-Type: application/json',
            'User-Agent: TigerEx-PHP-SDK/1.0.0',
        ];

        if (!empty($this->apiKey)) {
            $headers[] = 'X-MEX-APIKEY: ' . $this->apiKey;
        }

        if ($signed) {
            $signature = $this->sign($queryString . '&timestamp=' . $timestamp);
            $headers[] = 'X-MEX-SIGNATURE: ' . $signature;
        }

        $ch = curl_init();
        curl_setopt($ch, CURLOPT_URL, $url);
        curl_setopt($ch, CURLOPT_CUSTOMREQUEST, $method);
        curl_setopt($ch, CURLOPT_HTTPHEADER, $headers);
        curl_setopt($ch, CURLOPT_RETURNTRANSFER, true);
        curl_setopt($ch, CURLOPT_TIMEOUT_MS, $this->timeout);

        if ($method === 'POST' || $method === 'PUT') {
            curl_setopt($ch, CURLOPT_POSTFIELDS, json_encode($params));
        }

        $response = curl_exec($ch);
        $httpCode = curl_getinfo($ch, CURLINFO_HTTP_CODE);
        $error = curl_error($ch);
        curl_close($ch);

        if ($error) {
            throw new \Exception('cURL Error: ' . $error);
        }

        return json_decode($response, true);
    }

    // =========================================================================
    // MARKET DATA
    // =========================================================================

    public function ping(): array
    {
        return $this->request('GET', '/api/v3/ping');
    }

    public function time(): array
    {
        return $this->request('GET', '/api/v3/time');
    }

    public function exchangeInfo(string $symbol = ''): array
    {
        return $this->request('GET', '/api/v3/exchangeInfo', [
            'symbol' => $symbol,
        ]);
    }

    public function tickerPrice(string $symbol): array
    {
        return $this->request('GET', '/api/v3/ticker/price', [
            'symbol' => $symbol,
        ]);
    }

    public function ticker24h(string $symbol = ''): array
    {
        return $this->request('GET', '/api/v3/ticker/24hr', [
            'symbol' => $symbol,
        ]);
    }

    public function bookTicker(string $symbol): array
    {
        return $this->request('GET', '/api/v3/ticker/bookTicker', [
            'symbol' => $symbol,
        ]);
    }

    public function depth(string $symbol, int $limit = 100): array
    {
        return $this->request('GET', '/api/v3/depth', [
            'symbol' => $symbol,
            'limit' => $limit,
        ]);
    }

    public function trades(string $symbol, int $limit = 500): array
    {
        return $this->request('GET', '/api/v3/trades', [
            'symbol' => $symbol,
            'limit' => $limit,
        ]);
    }

    public function klines(string $symbol, string $interval = '1m', int $limit = 500): array
    {
        return $this->request('GET', '/api/v3/klines', [
            'symbol' => $symbol,
            'interval' => $interval,
            'limit' => $limit,
        ]);
    }

    public function avgPrice(string $symbol, int $minutes = 5): array
    {
        return $this->request('GET', '/api/v3/avgPrice', [
            'symbol' => $symbol,
            'minutes' => $minutes,
        ]);
    }

    // =========================================================================
    // ACCOUNT
    // =========================================================================

    public function account(): array
    {
        return $this->request('GET', '/api/v3/account', [], true);
    }

    public function myTrades(string $symbol, array $options = []): array
    {
        $params = array_merge(['symbol' => $symbol], $options);
        return $this->request('GET', '/api/v3/myTrades', $params, true);
    }

    // =========================================================================
    // ORDERS
    // =========================================================================

    public function order(string $symbol, int $orderId = null, string $origClientOrderId = null): array
    {
        $params = ['symbol' => $symbol];
        if ($orderId) $params['orderId'] = $orderId;
        if ($origClientOrderId) $params['origClientOrderId'] = $origClientOrderId;
        return $this->request('GET', '/api/v3/order', $params, true);
    }

    public function openOrders(string $symbol = ''): array
    {
        return $this->request('GET', '/api/v3/openOrders', [
            'symbol' => $symbol,
        ], true);
    }

    public function allOrders(string $symbol, array $options = []): array
    {
        $params = array_merge(['symbol' => $symbol], $options);
        return $this->request('GET', '/api/v3/allOrders', $params, true);
    }

    public function createOrder(array $order): array
    {
        return $this->request('POST', '/api/v3/order', $order, true);
    }

    public function cancelOrder(string $symbol, int $orderId = null, string $origClientOrderId = null): array
    {
        $params = ['symbol' => $symbol];
        if ($orderId) $params['orderId'] = $orderId;
        if ($origClientOrderId) $params['origClientOrderId'] = $origClientOrderId;
        return $this->request('DELETE', '/api/v3/order', $params, true);
    }

    // =========================================================================
    // CONVENIENCE METHODS
    // =========================================================================

    public function buyLimit(string $symbol, float $quantity, float $price, array $options = []): array
    {
        return $this->createOrder(array_merge([
            'symbol' => $symbol,
            'side' => 'BUY',
            'type' => 'LIMIT',
            'quantity' => $quantity,
            'price' => $price,
        ], $options));
    }

    public function sellLimit(string $symbol, float $quantity, float $price, array $options = []): array
    {
        return $this->createOrder(array_merge([
            'symbol' => $symbol,
            'side' => 'SELL',
            'type' => 'LIMIT',
            'quantity' => $quantity,
            'price' => $price,
        ], $options));
    }

    public function buyMarket(string $symbol, float $quantity, array $options = []): array
    {
        return $this->createOrder(array_merge([
            'symbol' => $symbol,
            'side' => 'BUY',
            'type' => 'MARKET',
            'quantity' => $quantity,
        ], $options));
    }

    public function sellMarket(string $symbol, float $quantity, array $options = []): array
    {
        return $this->createOrder(array_merge([
            'symbol' => $symbol,
            'side' => 'SELL',
            'type' => 'MARKET',
            'quantity' => $quantity,
        ], $options));
    }

    // =========================================================================
    // WALLET
    // =========================================================================

    public function depositAddress(string $coin, string $network = ''): array
    {
        $params = ['coin' => $coin];
        if ($network) $params['network'] = $network;
        return $this->request('GET', '/api/v3/deposit/address', $params, true);
    }

    public function depositHistory(array $options = []): array
    {
        return $this->request('GET', '/api/v3/deposit/history', $options, true);
    }

    public function withdraw(string $coin, string $address, float $amount, string $network = ''): array
    {
        $params = [
            'coin' => $coin,
            'address' => $address,
            'amount' => $amount,
        ];
        if ($network) $params['network'] = $network;
        return $this->request('POST', '/api/v3/withdraw/apply', $params, true);
    }

    public function withdrawHistory(array $options = []): array
    {
        return $this->request('GET', '/api/v3/withdraw/history', $options, true);
    }

    // =========================================================================
    // MARGIN
    // =========================================================================

    public function marginAccount(): array
    {
        return $this->request('GET', '/sapi/v3/margin/account', [], true);
    }

    public function createMarginOrder(array $order): array
    {
        return $this->request('POST', '/sapi/v3/margin/order', $order, true);
    }

    // =========================================================================
    // FUTURES
    // =========================================================================

    public function futuresAccount(): array
    {
        return $this->request('GET', '/fapi/v3/account', [], true);
    }

    public function futuresPosition(string $symbol = ''): array
    {
        return $this->request('GET', '/fapi/v3/position', [
            'symbol' => $symbol,
        ], true);
    }

    public function createFuturesOrder(array $order): array
    {
        return $this->request('POST', '/fapi/v3/order', $order, true);
    }

    // =========================================================================
    // STAKING
    // =========================================================================

    public function stakingStakes(string $product): array
    {
        return $this->request('GET', '/sapi/v1/staking/position', [
            'product' => $product,
        ], true);
    }

    public function stakingSubscribe(string $product, float $amount, string $asset): array
    {
        return $this->request('POST', '/sapi/v1/staking/subscribe', [
            'product' => $product,
            'amount' => $amount,
            'asset' => $asset,
        ], true);
    }

    public function stakingRedeem(string $product, float $amount, string $asset): array
    {
        return $this->request('POST', '/sapi/v1/staking/redeem', [
            'product' => $product,
            'amount' => $amount,
            'asset' => $asset,
        ], true);
    }
}

// ============================================================================
// WEBSOCKET CLIENT
// ============================================================================

class WebSocketClient
{
    private $ws;
    private $url;
    private $subscriptions = [];
    private $connected = false;

    public function __construct(bool $testnet = false, string $apiKey = '')
    {
        $this->url = ($testnet ? 'wss://stream-test.tigerex.com/ws' : 'wss://stream.tigerex.com/ws');
        if ($apiKey) {
            $this->url .= '?apiKey=' . $apiKey;
        }
    }

    public function connect(): bool
    {
        $this->ws = @fsockopen(
            parse_url($this->url, PHP_URL_HOST),
            parse_url($this->url, PHP_URL_PORT) ?: 443,
            $errno,
            $errstr,
            5
        );

        if (!$this->ws) {
            return false;
        }

        stream_set_blocking($this->ws, false);
        $this->connected = true;
        return true;
    }

    public function disconnect(): void
    {
        if ($this->ws) {
            fclose($this->ws);
            $this->ws = null;
        }
        $this->connected = false;
    }

    public function subscribe(array $streams): void
    {
        $this->send([
            'method' => 'SUBSCRIBE',
            'params' => $streams,
            'id' => time(),
        ]);
        $this->subscriptions = array_merge($this->subscriptions, $streams);
    }

    public function unsubscribe(array $streams): void
    {
        $this->send([
            'method' => 'UNSUBSCRIBE',
            'params' => $streams,
            'id' => time(),
        ]);
        $this->subscriptions = array_diff($this->subscriptions, $streams);
    }

    private function send(array $data): void
    {
        if ($this->ws) {
            fwrite($this->ws, json_encode($data) . "\n");
        }
    }

    public function isConnected(): bool
    {
        return $this->connected;
    }
}

// ============================================================================
// FACTORY
// ============================================================================

function createClient(
    string $apiKey = '',
    string $apiSecret = '',
    bool $testnet = false
): TigerExClient {
    return new TigerExClient($apiKey, $apiSecret, $testnet);
}

function createWebSocketClient(bool $testnet = false, string $apiKey = ''): WebSocketClient
{
    return new WebSocketClient($testnet, $apiKey);
}

// ============================================================================
// VERSION
// ============================================================================

const VERSION = '1.0.0';

File: sdks/java/TigerExClient.java
package com.tigerex.sdk;

import java.io.*;
import java.net.HttpURLConnection;
import java.net.URL;
import java.nio.charset.StandardCharsets;
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;
import java.util.*;
import javax.crypto.Mac;
import javax.crypto.spec.SecretKeySpec;

/**
 * TigerEx Java SDK
 * Production-grade Java SDK for TigerEx exchange API
 */
public class TigerExClient {
    
    private final String apiKey;
    private final String apiSecret;
    private final String baseUrl;
    private final int timeout;
    
    public TigerExClient(String apiKey, String apiSecret) {
        this(apiKey, apiSecret, false, 30000);
    }
    
    public TigerExClient(String apiKey, String apiSecret, boolean testnet, int timeout) {
        this.apiKey = apiKey;
        this.apiSecret = apiSecret;
        this.baseUrl = testnet ? "https://api-test.tigerex.com" : "https://api.tigerex.com";
        this.timeout = timeout;
    }
    
    // =========================================================================
    // PRIVATE METHODS
    // =========================================================================
    
    private String sign(String message) throws Exception {
        Mac mac = Mac.getInstance("HmacSHA256");
        SecretKeySpec secretKey = new SecretKeySpec(apiSecret.getBytes(StandardCharsets.UTF_8), "HmacSHA256");
        mac.init(secretKey);
        byte[] hash = mac.doFinal(message.getBytes(StandardCharsets.UTF_8));
        return bytesToHex(hash);
    }
    
    private String bytesToHex(byte[] bytes) {
        StringBuilder sb = new StringBuilder();
        for (byte b : bytes) {
            sb.append(String.format("%02x", b));
        }
        return sb.toString();
    }
    
    private String buildQueryString(Map<String, String> params) {
        List<String> keys = new ArrayList<>(params.keySet());
        Collections.sort(keys);
        StringBuilder sb = new StringBuilder();
        for (int i = 0; i < keys.size(); i++) {
            String key = keys.get(i);
            if (i > 0) sb.append("&");
            sb.append(key).append("=").append(params.get(key));
        }
        return sb.toString();
    }
    
    private String request(String method, String endpoint, Map<String, String> params, boolean signed) throws Exception {
        long timestamp = System.currentTimeMillis();
        String queryString = buildQueryString(params);
        
        String url = baseUrl + endpoint;
        if (method.equals("GET") && !queryString.isEmpty()) {
            url += "?" + queryString;
        }
        
        HttpURLConnection conn = (HttpURLConnection) new URL(url).openConnection();
        conn.setRequestMethod(method);
        conn.setRequestProperty("Content-Type", "application/json");
        conn.setRequestProperty("User-Agent", "TigerEx-Java-SDK/1.0.0");
        
        if (!apiKey.isEmpty()) {
            conn.setRequestProperty("X-MEX-APIKEY", apiKey);
        }
        
        if (signed) {
            String signature = sign(queryString + "&timestamp=" + timestamp);
            conn.setRequestProperty("X-MEX-SIGNATURE", signature);
        }
        
        if (method.equals("POST") || method.equals("PUT")) {
            conn.setDoOutput(true);
            try (OutputStream os = conn.getOutputStream()) {
                os.write(queryString.getBytes(StandardCharsets.UTF_8));
            }
        }
        
        int responseCode = conn.getResponseCode();
        BufferedReader reader = new BufferedReader(
            new InputStreamReader(
                responseCode >= 200 && responseCode < 300 
                    ? conn.getInputStream() 
                    : conn.getErrorStream(),
                StandardCharsets.UTF_8
            )
        );
        
        StringBuilder response = new StringBuilder();
        String line;
        while ((line = reader.readLine()) != null) {
            response.append(line);
        }
        reader.close();
        
        return response.toString();
    }
    
    // =========================================================================
    // MARKET DATA
    // =========================================================================
    
    public String ping() throws Exception {
        return request("GET", "/api/v3/ping", new HashMap<>(), false);
    }
    
    public String time() throws Exception {
        return request("GET", "/api/v3/time", new HashMap<>(), false);
    }
    
    public String exchangeInfo(String symbol) throws Exception {
        Map<String, String> params = new HashMap<>();
        if (symbol != null && !symbol.isEmpty()) {
            params.put("symbol", symbol);
        }
        return request("GET", "/api/v3/exchangeInfo", params, false);
    }
    
    public String tickerPrice(String symbol) throws Exception {
        return request("GET", "/api/v3/ticker/price", Map.of("symbol", symbol), false);
    }
    
    public String ticker24h(String symbol) throws Exception {
        Map<String, String> params = new HashMap<>();
        if (symbol != null && !symbol.isEmpty()) {
            params.put("symbol", symbol);
        }
        return request("GET", "/api/v3/ticker/24hr", params, false);
    }
    
    public String bookTicker(String symbol) throws Exception {
        return request("GET", "/api/v3/ticker/bookTicker", Map.of("symbol", symbol), false);
    }
    
    public String depth(String symbol, int limit) throws Exception {
        return request("GET", "/api/v3/depth", Map.of("symbol", symbol, "limit", String.valueOf(limit)), false);
    }
    
    public String trades(String symbol, int limit) throws Exception {
        return request("GET", "/api/v3/trades", Map.of("symbol", symbol, "limit", String.valueOf(limit)), false);
    }
    
    public String klines(String symbol, String interval, int limit) throws Exception {
        return request("GET", "/api/v3/klines", Map.of(
            "symbol", symbol,
            "interval", interval,
            "limit", String.valueOf(limit)
        ), false);
    }
    
    public String avgPrice(String symbol, int minutes) throws Exception {
        return request("GET", "/api/v3/avgPrice", Map.of("symbol", symbol, "minutes", String.valueOf(minutes)), false);
    }
    
    // =========================================================================
    // ACCOUNT
    // =========================================================================
    
    public String account() throws Exception {
        return request("GET", "/api/v3/account", new HashMap<>(), true);
    }
    
    public String myTrades(String symbol, Map<String, String> options) throws Exception {
        Map<String, String> params = new HashMap<>();
        params.put("symbol", symbol);
        if (options != null) params.putAll(options);
        return request("GET", "/api/v3/myTrades", params, true);
    }
    
    // =========================================================================
    // ORDERS
    // =========================================================================
    
    public String order(String symbol, Long orderId, String origClientOrderId) throws Exception {
        Map<String, String> params = new HashMap<>();
        params.put("symbol", symbol);
        if (orderId != null) params.put("orderId", String.valueOf(orderId));
        if (origClientOrderId != null) params.put("origClientOrderId", origClientOrderId);
        return request("GET", "/api/v3/order", params, true);
    }
    
    public String openOrders(String symbol) throws Exception {
        Map<String, String> params = new HashMap<>();
        if (symbol != null && !symbol.isEmpty()) {
            params.put("symbol", symbol);
        }
        return request("GET", "/api/v3/openOrders", params, true);
    }
    
    public String allOrders(String symbol, Map<String, String> options) throws Exception {
        Map<String, String> params = new HashMap<>();
        params.put("symbol", symbol);
        if (options != null) params.putAll(options);
        return request("GET", "/api/v3/allOrders", params, true);
    }
    
    public String createOrder(Map<String, String> order) throws Exception {
        return request("POST", "/api/v3/order", order, true);
    }
    
    public String cancelOrder(String symbol, Long orderId, String origClientOrderId) throws Exception {
        Map<String, String> params = new HashMap<>();
        params.put("symbol", symbol);
        if (orderId != null) params.put("orderId", String.valueOf(orderId));
        if (origClientOrderId != null) params.put("origClientOrderId", origClientOrderId);
        return request("DELETE", "/api/v3/order", params, true);
    }
    
    // =========================================================================
    // CONVENIENCE METHODS
    // =========================================================================
    
    public String buyLimit(String symbol, double quantity, double price, Map<String, String> options) throws Exception {
        Map<String, String> order = new HashMap<>();
        order.put("symbol", symbol);
        order.put("side", "BUY");
        order.put("type", "LIMIT");
        order.put("quantity", String.valueOf(quantity));
        order.put("price", String.valueOf(price));
        if (options != null) order.putAll(options);
        return createOrder(order);
    }
    
    public String sellLimit(String symbol, double quantity, double price, Map<String, String> options) throws Exception {
        Map<String, String> order = new HashMap<>();
        order.put("symbol", symbol);
        order.put("side", "SELL");
        order.put("type", "LIMIT");
        order.put("quantity", String.valueOf(quantity));
        order.put("price", String.valueOf(price));
        if (options != null) order.putAll(options);
        return createOrder(order);
    }
    
    public String buyMarket(String symbol, double quantity, Map<String, String> options) throws Exception {
        Map<String, String> order = new HashMap<>();
        order.put("symbol", symbol);
        order.put("side", "BUY");
        order.put("type", "MARKET");
        order.put("quantity", String.valueOf(quantity));
        if (options != null) order.putAll(options);
        return createOrder(order);
    }
    
    public String sellMarket(String symbol, double quantity, Map<String, String> options) throws Exception {
        Map<String, String> order = new HashMap<>();
        order.put("symbol", symbol);
        order.put("side", "SELL");
        order.put("type", "MARKET");
        order.put("quantity", String.valueOf(quantity));
        if (options != null) order.putAll(options);
        return createOrder(order);
    }
    
    // =========================================================================
    // WALLET
    // =========================================================================
    
    public String depositAddress(String coin, String network) throws Exception {
        Map<String, String> params = new HashMap<>();
        params.put("coin", coin);
        if (network != null && !network.isEmpty()) {
            params.put("network", network);
        }
        return request("GET", "/api/v3/deposit/address", params, true);
    }
    
    public String depositHistory(Map<String, String> options) throws Exception {
        return request("GET", "/api/v3/deposit/history", options != null ? options : new HashMap<>(), true);
    }
    
    public String withdraw(String coin, String address, double amount, String network) throws Exception {
        Map<String, String> params = new HashMap<>();
        params.put("coin", coin);
        params.put("address", address);
        params.put("amount", String.valueOf(amount));
        if (network != null && !network.isEmpty()) {
            params.put("network", network);
        }
        return request("POST", "/api/v3/withdraw/apply", params, true);
    }
    
    public String withdrawHistory(Map<String, String> options) throws Exception {
        return request("GET", "/api/v3/withdraw/history", options != null ? options : new HashMap<>(), true);
    }
    
    // =========================================================================
    // MARGIN
    // =========================================================================
    
    public String marginAccount() throws Exception {
        return request("GET", "/sapi/v3/margin/account", new HashMap<>(), true);
    }
    
    public String createMarginOrder(Map<String, String> order) throws Exception {
        return request("POST", "/sapi/v3/margin/order", order, true);
    }
    
    // =========================================================================
    // FUTURES
    // =========================================================================
    
    public String futuresAccount() throws Exception {
        return request("GET", "/fapi/v3/account", new HashMap<>(), true);
    }
    
    public String futuresPosition(String symbol) throws Exception {
        Map<String, String> params = new HashMap<>();
        if (symbol != null && !symbol.isEmpty()) {
            params.put("symbol", symbol);
        }
        return request("GET", "/fapi/v3/position", params, true);
    }
    
    public String createFuturesOrder(Map<String, String> order) throws Exception {
        return request("POST", "/fapi/v3/order", order, true);
    }
    
    // =========================================================================
    // MAIN
    // =========================================================================
    
    public static void main(String[] args) {
        try {
            TigerExClient client = new TigerExClient("test_key", "test_secret", true);
            
            // Test ping
            System.out.println("Ping: " + client.ping());
            
            // Test time
            System.out.println("Time: " + client.time());
            
            // Test ticker
            System.out.println("Ticker: " + client.tickerPrice("BTCUSDT"));
            
        } catch (Exception e) {
            e.printStackTrace();
        }
    }
}